iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp

```diff
@@ -734,7 +742,77 @@ class RankReducedDotGeneral : public OpRewritePattern<mhlo::DotGeneralOp> {
 
     return success();
   }
-};  // namespace
+};
+
+// Generates Gaussian noise with uniform random generator based on Box-Muller
+// transform.
+class ExpandRngNormalToRngUniform : public OpRewritePattern<mhlo::RngNormalOp> {
+ public:
+  using OpRewritePattern<mhlo::RngNormalOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::RngNormalOp op,
+                                PatternRewriter &rewriter) const override {
+    auto resTy = op.getType().dyn_cast<RankedTensorType>();
+    // We can support static shapes, but it's easier to implement Box-Muller
+    // transform if we know the number of elements.
+    if (!resTy || !resTy.hasStaticShape()) return failure();
+
+    // The algorithm requires even numbers and will generate pairs.
+    auto numElems = resTy.getNumElements();
+    if (numElems & 1) numElems++;
+    auto halfNumElems = numElems / 2;
+
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+    Value epsilon = getF32SplatConst(b, /*shapes=*/{},
+                                     std::numeric_limits<float>::epsilon());
+    Value upperBound = getF32SplatConst(b, /*shapes=*/{}, 1.0);
+
+    // Create pairs of random numbers, make sure u1 is greater than epsilon.
+    auto shapeTensor =
+        b.create<mhlo::ConstOp>(make1DElementsAttr(rewriter, numElems));
+    auto rngType = RankedTensorType::get({numElems}, resTy.getElementType());
+    auto rng =
+        b.create<mhlo::RngUniformOp>(rngType, epsilon, upperBound, shapeTensor);
```

Comment:
High level question, these ops are in fact stateful how seed is passed around ?

Reply:
I did some research, and put some notes to https://github.com/google/iree/issues/5298.

Regarding the seeds, there are two seeds in tensorflow. One is shared (globally), and another is carried by the op. In IREE, we can add global seed at somewhere (need more exploration). The MHLO ops do not have seed information. The seed seems to be dropped during TF -> MHLO.

```
foreach srcDstOpPair = [[TF_RandomUniformOp, HLO_RngUniformOp],
                        [TF_RandomStandardNormalOp, HLO_RngNormalOp]] in {
// TODO(b/148269299): handle random number generator seeds/states correctly.
def : Pat<(srcDstOpPair[0]:$old $shape, $seed, $seed2),
          (srcDstOpPair[1]
            (HLO_ConstOp
              (NativeCodeCall<"$_builder.getFloatAttr(old.dtype(), 0.0)">)),
            (HLO_ConstOp
              (NativeCodeCall<"$_builder.getFloatAttr(old.dtype(), 1.0)">)),
            (CastValueToI64 $old, $shape)),
          [(IsShapedTensor $shape)]>;
}
```

We have stateless uniform rng_generator lowering (which hard-codes the seed) upstreamed. The PR aims to emulate normal distribution through uniform rng, so the stateful/stateless issue is not critical here.

To support stateful rng generator, we will have to create a flow variable, use it in a callback function when lowering rng_* ops.